generator/templates/server/autoconfigureapi.gotmpl

```diff
@@ -0,0 +1,149 @@
+)
+
+{{ with .GenOpts }}
+//go:generate swagger generate server --target {{ .TargetPath }} --name {{ .Name }} --spec {{ .SpecPath }}
+{{- if .APIPackage }}{{ if ne .APIPackage "operations" }} --api-package {{ .APIPackage }}{{ end }}{{ end }}
+{{- if .ModelPackage }}{{ if ne .ModelPackage "models" }} --model-package {{ .ModelPackage }}{{ end }}{{ end }}
+{{- if .ServerPackage }}{{ if ne .ServerPackage "restapi"}} --server-package {{ .ServerPackage }}{{ end }}{{ end }}
+{{- if .ClientPackage }}{{ if ne .ClientPackage "client" }} --client-package {{ .ClientPackage }}{{ end }}{{ end }}
+{{- if .ImplementationPackage }} --implementation-package {{ .ImplementationPackage }}{{ end }}
+{{- if .TemplateDir }} --template-dir {{ .TemplateDir }}{{ end }}
+{{- range .Operations }} --operation {{ . }}{{ end }}
+{{- range .Tags }} --tags {{ . }}{{ end }}
+{{- if .Principal }} --principal {{ .Principal }}{{ end }}
+{{- if .DefaultScheme }}{{ if ne .DefaultScheme "http" }} --default-scheme {{ .DefaultScheme }}{{ end }}{{ end }}
+{{- range .Models }} --model {{ . }}{{ end }}
+{{- if or (not .IncludeModel) (not .IncludeValidator) }} --skip-models{{ end }}
+{{- if or (not .IncludeHandler) (not .IncludeParameters ) (not .IncludeResponses) }} --skip-operations{{ end }}
+{{- if not .IncludeSupport }} --skip-support{{ end }}
+{{- if not .IncludeMain }} --exclude-main{{ end }}
+{{- if .ExcludeSpec }} --exclude-spec{{ end }}
+{{- if .DumpData }} --dump-data{{ end }}
+{{- if .StrictResponders }} --strict-responders{{ end }}
+{{ end }}
+
+// This file auto configures the api backend implementation to a single object or interface "Impl",
+// which must implement all methods to handle each api, and server configurations.
+// This file can be regenerated without losing the api server and backend wirings,
+// but requirement on how to organize backend implementation is more strict.
+var Impl = {{.ImplementationPackageAlias}}.New()
+
+func configureFlags(api *{{.APIPackageAlias}}.{{ pascalize .Name }}API) {
+  Impl.ConfigureFlags(api)
+}
+
+func configureAPI(api *{{.APIPackageAlias}}.{{ pascalize .Name }}API) http.Handler {
+
+  api.ServeError = errors.ServeError
+
+  api.UseSwaggerUI()
+
+  {{ range .Consumes }}
+    {{- if .Implementation }}
+  api.{{ pascalize .Name }}Consumer = {{ .Implementation }}
+    {{- else }}
+  api.{{ pascalize .Name }}Consumer = runtime.ConsumerFunc(func(r io.Reader, target interface{}) error {
+    return Impl.{{ pascalize .Name }}Consume(r, target)
+  })
+    {{- end }}
+  {{- end }}
+  {{ range .Produces }}
+    {{- if .Implementation }}
+  api.{{ pascalize .Name }}Producer = {{ .Implementation }}
+    {{- else }}
+  api.{{ pascalize .Name }}Producer = runtime.ProducerFunc(func(w io.Writer, data interface{}) error {
+    return Impl.{{ pascalize .Name }}Produce(w, target)
+  })
+    {{- end }}
+  {{- end}}
+  {{ range .SecurityDefinitions }}
+    {{- if .IsBasicAuth }}
+  // Applies when the Authorization header is set with the Basic scheme
```

Comment:
This is good, although I'd have lower requirements on the API implementation regarding the presence of all technical methods (Auth, ConfigureFlags, etc.).

This is how I would do it. Maybe this is something to consider for a later improvement. Feel free to comment on these ideas.

```go
type Implements{{ API name }}API interface {
  {{ ... all API handlers }} // attention! there is a CLI opt that changes the signature and requires specialized responders instead of the generic middleware.Responder
}

type Authable{{ API name}} interface {
   ... //  {{ ...}}Auth methods
}

type ConfigurableAPI interface {
... // ConfigureFlags, ConfigureServer, SetupMiddleware, etc...
   CustomConfigure(ImplementsSimpleToDoListAPIAPI)
}

func configureAPI(injected ImplementsSimpleToDoListAPIAPI) http.Handler {
 ...
 if authable, ok := injected.(Authable{{ API name}}) ; ok {
  api.{{ .ID }}Auth = authable.{{ .ID }}Auth
}
... etc you probably got the gist of it

```
This comment is not a blocker: we may proceed like this and improve it with another PR.

Reply:
I have some pending changes to displaying all the functions required for implementation obj in interfaces. I will push soon.

I thought about using interface assertion, and make this auto_configure_xxx.go easily compile, but the user should catch the unimplemented error in compile time, instead of runtime. If the interface assertion failed (in your authable) above, it is seems more appropriate to panic, instead of assigning a unimplemented runtime error.

Reply:
I added interfaces in the auto-configure template. If we think that type assertion on interfaces are preferred, I can add it in another PR.

Reply:
Just leave it as you first intended. Let's see how people react to this.